### PR TITLE
Ruby Group Capstone project

### PR DESCRIPTION
**Pull request by Goldinium**
- Each class from the [UML class diagram](https://raw.githubusercontent.com/microverseinc/curriculum-ruby/main/group-capstone/images/catalog_of_my_things.png?token=GHSAT0AAAAAACH5UDWDY64CIYLFEWBKND6KZLAS2OA) is created in a separate .rb file.
- All associations (1-to-many relationships) and aggregations (parent-child relationships) between classes should be implemented.

- All properties visible in the diagram are defined and set up in the constructor method. _Exception: properties for the 1-to-many relationships are NOT set in the constructor method. Instead, they have a custom setter method created._
- All methods visible in the diagram have been implemented:
    - add_item method (in all classes that have that method)
        -  takes an instance of the Item class as an input.
        -  add the input item to the collection of items.
        -  add self as a property of the item object (by using the correct setter from the item object).
    - can_be_archived?() in the Item class
        -  return true if published_date is older than 10 years.
        - otherwise, it return false.
    - move_to_archive() in the Item class
        -  reuse can_be_archived?() method.
        -  change the archived property to true if the result of the can_be_archived?() method is true.
        -  do nothing if the result of the can_be_archived?() method is false.
    - can_be_archived?() in the Book class
        -  override the method from the parent class.
        -  return true if parent's method returns true OR if cover_state equals to "bad".
        - otherwise, it return false.
    - can_be_archived?() in the MusicAlbum class
        -  override the method from the parent class.
        -  return true if parent's method returns true AND if on_spotify equals true.
        - otherwise, it return false.
    - can_be_archived?() in the Movie class
        - override the method from the parent class.
        - return true if parent's method returns true OR if silent equals true.
        - otherwise, it  return false.
    - can_be_archived?() in the Game class
        -  override the method from the parent class.
        -  return true if parent's method returns true AND if last_played_at is older than 2 years.
        - otherwise, it return false.
- Added unit tests for all implemented methods.

#### User interface

- Created a main.rb file that will serve as your console app entry-point.
- The console app, at the start:
    - Present the user with a list of options to perform.
    - Let users choose an option.
    - If needed, ask for parameters for the option.
    - Have a way to quit the app.
- The following options are available:
    - List all books
    - List all music albums
    - List all movies
    - List of games
    - List all genres (e.g 'Comedy', 'Thriller')
    - List all labels (e.g. 'Gift', 'New')
    - List all authors (e.g. 'Stephen King')
    - List all sources (e.g. 'From a friend', 'Online shop')
    - Add a book
    - Add a music album
    - Add a movie
    - Add a game
 - All data is preserved by saving collections in .json files.

### Database
- Created a schema.sql file with tables that will are analogical to the structure of the classes in the app. - added all properties and associations from the parent Item class as table columns to all tables based on the child classes.

### CODE REVIEW

**Review by VICTOR-CHUKWU — CHANGES_REQUESTED** [review]
Hi @Goldinium  and @Al-Amin-Khan-Shakil  and @lucy-sees  and @evansnyamekye , 
## STATUS: CHANGES REQUIRED :recycle: 

Good job so far!
There are some issues that you still need to work on to prepare your project for the final evaluation but you are almost there!

## Suggested changes
_Check the comments under the review._

You can use as many of my suggestions as you want. If there is anything you would like to skip - feel free to do that. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

**Comment by Goldinium** [comment]
@VICTOR-CHUKWU  the app crashing due to invalid date input and all edge-cases are not part of the project requirements, I don't understand why you are mentioning them

Can you explain?

File: classes/item.rb [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+  def move_to_archive
+    @archived = true if can_be_archived?
+  end
+
+  private
+
+  def can_be_archived?
+    Date.today.year - @publish_date.year > 10
+  end
```
[the remark this answers is not in the document]
Comment (Al-Amin-Khan-Shakil) [reply classes/item.rb:42]: Bro this Item class is a requirement in or group task and the group task does not require any test. It is not instructed in our requirements that's why we didn't implement it.

**Comment by Goldinium** [comment]
We added some methods to prevent the app crashing and ask for a correct date if an invalid one is provided by the user.

File: modules/game_module.rb [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        \t\t| #{game.last_played_at} \t\t|"
+        puts '-------------------------------------------------------------------------'
+      end
+    end
+  end
+
+  def add_game()
+    puts "\nAdd a game:"
+    puts 'Published date (dd/mm/yy): '
+    publish_date = Date.parse(gets.chomp)
```
[the remark this answers is not in the document]
Comment (Abenezer-Tilahun) [reply modules/game_module.rb]: Fixed.

File: modules/game_module.rb [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    print 'Multiplayer [Y/N]: '
+    multiplayer = gets.chomp.downcase == 'y'
+    print 'Last played at (dd/mm/yy): '
+    last_played_at = Date.parse(gets.chomp)
```
[the remark this answers is not in the document]
Comment (Abenezer-Tilahun) [reply modules/game_module.rb]: Fixed.

File: app.rb [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+  end
+
+  def add_new_album
+    puts 'Enter a date: DD/MM/YYYY'
+    album_date = gets.chomp
+    puts 'enter Y if it\'s on spotify or N if it\'s not'
+    on_spotify = gets.chomp.downcase == 'y'
+    puts 'Enter a genre:'
+    genre = gets.chomp
+    new_album = MusicAlbum.new(on_spotify, album_date)
```
[the remark this answers is not in the document]
Comment (Abenezer-Tilahun) [reply app.rb]: Fixed.

File: app.rb [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+    puts 'Source-name: '
+    name = gets.chomp
+    new_movie = Movie.new(silet, publish_date)
+    new_source = Source.new(name)
+    new_source.add_item(new_movie)
+    @movies << new_movie
+    @sources << new_source
+    DataManager.save_movie(@movies)
+    DataManager.save_source(@sources)
+    puts 'Book added successfully'
```
[the remark this answers is not in the document]
Comment (Abenezer-Tilahun) [reply app.rb]: Fixed.

File: classes/item.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (Abenezer-Tilahun) [reply classes/item.rb:42]: Fixed.

**Review by Abenezer-Tilahun — APPROVED** [review]
Hi @Goldinium 
 
Your project is complete  👍 💯   There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights 👏🏻

✔️ No linter errors.
✔️ You used correct Git-Flow.
✔️ Professional README  
✔️ Good Commit Messages
✔️ Good job for Implemented Suggested changes.


## Optional suggestions 

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application.

**Pull request merged**